bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java

```diff
@@ -0,0 +1,310 @@
+        .setIsEnabled(bulkOptions.useBulkApi())
+        .setElementCountThreshold(Long.valueOf(bulkOptions.getBulkMaxRowKeyCount()))
+        .setRequestByteThreshold(bulkOptions.getBulkMaxRequestSize())
+        .setFlowControlSettings(flowControlBuilder.build());
+
+    // TODO: implement bulkMutationThrottling & bulkMutationRpcTargetMs, once available
+    if (options.getRetryOptions().enableRetries()) {
+      builder.bulkMutationsSettings()
+          .setBatchingSettings(batchSettingsBuilder.build())
+          .setRetryableCodes(DEFAULT_RETRY_CODES)
+          .setRetrySettings(defaultRetrySettings(options));
+    } else {
+      builder.bulkMutationsSettings()
+          .setSimpleTimeoutNoRetries(DEFAULT_RPC_TIMEOUT)
+          .getRetryableCodes().clear();
+    }
+  }
+
+  /**
+   * To build sampleRowKey with default Settings based on retries setting.
+   *
+   * @param builder a {@link BigtableDataSettings.Builder} object.
+   * @param bulkMutation a {@link BulkOptions} object.
+   */
+  private static void buildSampleRowKeys(Builder builder, BigtableOptions options) {
+    if (options.getRetryOptions().enableRetries()) {
+      builder.sampleRowKeysSettings()
+          .setRetryableCodes(DEFAULT_RETRY_CODES)
+          .setRetrySettings(defaultRetrySettings(options));
+    } else {
+      builder.sampleRowKeysSettings()
+          .setSimpleTimeoutNoRetries(DEFAULT_RPC_TIMEOUT)
+          .getRetryableCodes().clear();
+    }
+  }
+
+  /**
+   * To build mutateRowSettings with default Settings based on retries setting.
+   *
+   * @param builder a {@link BigtableDataSettings.Builder} object.
+   * @param bulkMutation a {@link BulkOptions} object.
+   */
+  private static void buildMutateRow(Builder builder, BigtableOptions options) {
+    if (options.getRetryOptions().enableRetries()) {
+      builder.mutateRowSettings()
+          .setRetryableCodes(DEFAULT_RETRY_CODES)
+          .setRetrySettings(defaultRetrySettings(options));
+    } else {
+      builder.sampleRowKeysSettings()
+          .setSimpleTimeoutNoRetries(DEFAULT_RPC_TIMEOUT)
+          .getRetryableCodes().clear();
+    }
+  }
+
+  /**
+   * To build readRows with default Settings based on retries setting.
+   *
+   * @param builder a {@link BigtableDataSettings.Builder} object.
+   * @param bulkMutation a {@link BulkOptions} object.
+   */
+  private static void buildReadRows(Builder builder, BigtableOptions options) {
+    // TODO: set readPartialRowTimeout for watchdog timer, taken BigtableSession#setupWatchdog()
+    if(options.getRetryOptions().enableRetries()) {
+      builder.readRowsSettings()
+          .setIdleTimeout(ofMillis(options.getRetryOptions().getReadPartialRowTimeoutMillis()))
+          .setRetryableCodes(DEFAULT_RETRY_CODES)
+          .setRetrySettings(defaultRetrySettings(options));
+    } else {
+      builder.readRowsSettings()
+          .setIdleTimeout(ofMillis(options.getRetryOptions().getReadPartialRowTimeoutMillis()))
```

Comment:
This is incorrect idleTimeout is the delay for the user leaving a stream open not the timeout for the server. You want to set the streamingWaitTimeout. There are 2 ways to do it:
1. the rpcTimeout of the RetrySettings gets translated into a waitTimeout
2. set it in calloptions

1 will work for when retries are enabled. However when retries are disabled, you would have to augment the calloptions at the callsite, which is awkward. Please open a ticket in gax-java to expose waitTimeout at the callable level

Reply:
I would open an issue stating this above scenario. Thanks